### ISSUE TITLE
"prefer-function-over-method" should allow methods that are required to satisfy an interface

### ISSUE THREAD

**Issue by treythomas123** [palantir/tslint#3564]
### Bug Report

- __TSLint version__: 5.7.0
- __TypeScript version__: 2.4.2
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
interface SomeInterface {
  someMethod: () => void;
}

class Impl implements SomeInterface {
  someMethod(): void {
    return;
  }
}
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "prefer-function-over-method": true
    }
}
```

#### Actual behavior
```ERROR: test.ts[6, 3]: Class method does not use 'this'. Use a function instead.```

#### Expected behavior
TSLint should pass, because extracting this method into a function would cause the class to no longer satisfy the interface.

I know you can add the "allow-public" config, which would allow this to pass, but I want to catch other public methods that aren't required by interfaces.

**Comment by aervin** [comment]
Will this pass?

```ts
interface ISomeInterface {
    doSomething: () => void;
}

export class SomeClass implements ISomeInterface {
    public doSomething: () => void = foo;
}

export function foo(): void {
    // ...stuff
    return;
}
```

Doesn't seem too bad. And the rule in question is for those who want to use the `function` syntax _whenever_ possible after all.

**Comment by sbj42** [comment]
I have run into this as well.  I think the workaround in the previous comment is clever, but my preference for overrides (both public and protected) is to keep them in the class.  I think the aliasing technique makes the code a little harder to read.  I use the rule to detect helper methods that should be functions, not "all methods that could be functions".

So, my preference would be to make the change as requested by @treythomas123.  If there are people who prefer the aliasing technique, then an acceptable plan B would be to add an `allow-override` config option.

**Comment by artola** [comment]
`allow-override` is a great idea

**Comment by sbj42** [comment]
@ajafff I'm interested in your opinion here.  You wrote this line in `preferFunctionOverMethodRule.ts`:
~~~ts
// TODO: handle the override keyword once it lands in the language
~~~
Is it necessary to wait for the override keyword - is it possible to detect a matching signature in a parent class or interface?  If so, should such methods always be exempt, or should a config option be necessary?

**Comment by rope-hmg** [comment]
Any update on this? My current work around is to use `allow-public`, but I would rather this.

**Comment by JoshuaKGoldberg** [comment]
@rope-hmg no update other than #4534. This issue will be closed soon if no PR is sent to address it. If you have the time, that'd be swell!

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
